Automattic/Gravatar-SDK-Android
# Initial code for the QuickEditor that should start with the OAuth

### Description
~I am opening this as a Draft to discuss the API and the structure in general. The code is not final yet.~

This is the module structure for now:
<img width="449" alt="Screenshot 2024-07-24 at 12 47 47" src="https://github.com/user-attachments/assets/82ffcc84-e0f9-4610-a26b-7ce4dadbb0b1">


- `oauth` package will hold all the OAuth related code (UI and network)
-  `editor` will contain what's related to the Editor including the possible error and result updates


**Integration**

Jetpack Compose
```
GravatarQuickEditorBottomSheet(
            gravatarQuickEditorParams = GravatarQuickEditorParams {
                appName = "yourAppName"
                oAuthParams = OAuthParams {
                    clientId = "clientId
                    clientSecret = "clientSecret"
                    redirectUri = "redirectUri"
                }
                scope = GravatarQuickEditorScope.AVATARS
            },
            onDismiss = {            },
   )
```

Activity/Fragment 
```
GravatarQuickEditor.show(
                activity = this,
                gravatarQuickEditorParams = GravatarQuickEditorParams {
                    appName = "yourAppName"
                    oAuthParams = OAuthParams {
                        clientId = "clientId"
                        clientSecret = "clientSecret"
                        redirectUri = "redirectUri"
                    }
                    scope = GravatarQuickEditorScope.AVATARS
                },
                onDismiss = {                },
            )
)
```


For the `OAuthParams` class I've used the same pattern as we use for the classes generated by OpenAPI - meaning the builder for Java and DSL for Kotlin. 

~About the `scope = GravatarQuickEditorScope.AVATARS` - I've followed the web here - they offer different scopes to edit by the user. Not sure if we are planning to have more than just Avatars so we could drop it for now.~


`I will continue with cleaning up and extracting the `oauth` package. We need ViewModel and separate services but let me know what you think so far. ` - this is handled in a separate [PR](https://github.com/Automattic/Gravatar-SDK-Android/pull/240)

### Testing Steps


Add these to your `local.properties` 
```
demo-app.wordpress.oauth.clientId=yourId
demo-app.wordpress.oauth.clientSecret=yourSecret
demo-app.wordpress.oauth.redirectUri=yourURI
```

and play with the demo app. 

1. Tap the `UpdateAvatar`
2. Continue with the required steps to finish the OAuthFlow
3. `Insert the real avatar picker page` text should appear - this is a placeholder for the avatar picker UI
4. Tap the above text
5. Toast message should appear with a confirmation
6. Tap the `UpdateAvatar` 
7. Close the browser with the `x` button
8. Tap the `Done` button
9. Toast with `Dissmised` info is shown
10. Tap the `UpdateAvatar` 
11. Tap `Deny`
12. Toast with `Error` info is shown
